### PR TITLE
Avoid crash on Linux when opening many geopackages at once (fixes #43620)

### LINKED ISSUE #43620: QGIS on linux crashes when opening many geopackages at once
**Describe the bug**
Drag-n-dropping 200 geopackages at once in QGIS causes a crash:
```
../src/core/providers/ogr/qgsogrprovider.cpp:5250 : (open) [26026ms] OGR failed to opened in update mode, trying in read-only mode
../src/core/providers/ogr/qgsogrprovider.cpp:5250 : (open) [8ms] OGR failed to opened in update mode, trying in read-only mode
../src/core/providers/ogr/qgsogrprovider.cpp:6188 : (getLayer) [0ms] Cannot reuse existing opened dataset(s) on /home/uclaros/work/kt/aitol/agnik/aitol_20200824.gpkg since it has been modified
../src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2021-06-08T12:32:30 OGR[1] Cannot open /home/uclaros/work/kt/aitol/agnik/aitol_20200824.gpkg.(sqlite3_open(/home/uclaros/work/kt/aitol/agnik/aitol_20200824.gpkg) failed: unable to open database file)
../src/core/providers/ogr/qgsogrprovider.cpp:5250 : (open) [81ms] OGR failed to opened in update mode, trying in read-only mode
../src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2021-06-08T12:32:30 OGR[1] Cannot open /home/uclaros/work/kt/aitol/agnik/aitol_20200825.gpkg.()
ERROR: Status 2: File /home/uclaros/work/kt/aitol/agnik/aitol_20200825.gpkg could not be found
ERROR: Status 2: File /home/uclaros/work/kt/aitol/agnik/aitol_20200825.gpkg could not be found
../src/core/qgsmessagelog.cpp:29 : (logMessage) [1ms] 2021-06-08T12:32:30 [1] Unable to load /home/uclaros/work/kt/aitol/agnik/aitol_20200825.gpkg
../src/core/qgsmessagelog.cpp:29 : (logMessage) [9ms] 2021-06-08T12:32:30 Messages[2] Invalid Data Source : /home/uclaros/work/kt/aitol/agnik/aitol_20200825.gpkg is not a valid or recognized data source.

(qgis:1407617): GLib-ERROR **: 12:32:31.383: Creating pipes for GWakeup: Too many open files
QGIS died on signal 5[New LWP 1407618]
[New LWP 1407619]
[New LWP 1407625]
[New LWP 1407626]
[New LWP 1407627]
[New LWP 1407628]
[New LWP 1407629]
[New LWP 1407630]
[New LWP 1407634]
[New LWP 1407641]
[New LWP 1407700]
[New LWP 1407701]
[New LWP 1407702]
[New LWP 1407749]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007ff27c7743ff in __GI___poll (fds=0x55570520d8e0, nfds=7, timeout=182) at ../sysdeps/unix/sysv/linux/poll.c:29
29      ../sysdeps/unix/sysv/linux/poll.c: No such file or directory.
[Current thread is 1 (Thread 0x7ff26c1d71c0 (LWP 1407617))]
#0  0x00007ff27c7743ff in __GI___poll (fds=0x55570520d8e0, nfds=7, timeout=182) at ../sysdeps/unix/sysv/linux/poll.c:29
        resultvar = 18446744073709551100
        sc_cancel_oldtype = 0
#1  0x00007ff274ecd0ae in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007ff274ecd1cf in g_main_context_iteration () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#3  0x00007ff27f3f14bf in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#4  0x00007ff27f39892b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#5  0x00007ff27f3a0ba0 in QCoreApplication::exec() () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#6  0x00005557031f79e3 in main(int, char**) (argc=1, argv=0x7ffe904ea5d8) at ../src/app/main.cpp:1644
        preApplicationLogMessages = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = {<std::__atomic_base<int>> = {static _S_alignment = 4, _M_i = -1}, static is_always_lock_free = true}}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x55570506e150}, d = 0x55570506e150}}, <No data fields>}
        __FUNCTION__ = "main"
        mySnapshotFileName = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}
        configLocalStorageLocation = {static null = {<No data fields>}, d = 0x555705078a60}
        profileName = {static null = {<No data fields>}, d = 0x555705076ee0}
        mySnapshotWidth = 800
        mySnapshotHeight = 600
        myHideSplash = false
        settingsMigrationForce = false
        mySkipVersionCheck = false
        hideBrowser = false
        myRestoreDefaultWindowState = false
        myRestorePlugins = true
        myCustomization = true
        dxfOutputFile = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}
        dxfSymbologyMode = QgsDxfExport::SymbolLayerSymbology
        dxfScale = 50000
        dxfEncoding = {static null = {<No data fields>}, d = 0x555703203d00 <main::{lambda()#2}::operator()() const::qstring_literal>}
        dxfMapTheme = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}
        dxfExtent = {mXmin = 0, mYmin = 0, mXmax = 0, mYmax = 0}
        takeScreenShots = false
        screenShotsPath = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}
        screenShotsCategories = 0
        myInitialExtent = {static null = {<No data fields>}, d = 0x555703203d40 <main::{lambda()#3}::operator()() const::qstring_literal>}
        translationCode = {static null = {<No data fields>}, d = 0x5557050abab0}
        authdbdirectory = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}
        pythonfile = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}
        customizationfile = {static null = {<No data fields>}, d = 0x5557059f3260}
        globalsettingsfile = {static null = {<No data fields>}, d = 0x55570506e7e0}
        openClProgramFolder = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}
        args = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = {<std::__atomic_base<int>> = {static _S_alignment = 4, _M_i = -1}, static is_always_lock_free = true}}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x55570506e3f0}, d = 0x55570506e3f0}}, <No data fields>}
        myUseGuiFlag = true
        rootProfileFolder = {static null = {<No data fields>}, d = 0x555705078b20}
        manager = {<QObject> = {<No data fields>}, static staticMetaObject = {d = {superdata = {direct = 0x7ff27f626100 <QObject::staticMetaObject>}, stringdata = 0x7ff27e2a91a0 <qt_meta_stringdata_QgsUserProfileManager>, data = 0x7ff27e2a9220 <qt_meta_data_QgsUserProfileManager>, static_metacall = 0x7ff27d130f80 <QgsUserProfileManager::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, mWatchProfiles = false, mWatcher = std::unique_ptr<class QFileSystemWatcher> = {get() = 0x0}, mRootProfilePath = {static null = {<No data fields>}, d = 0x555705078b20}, mUserProfile = std::unique_ptr<class QgsUserProfile> = {get() = 0x0}, mSettings = std::unique_ptr<class QSettings> = {get() = 0x55570506e2d0}}
        profile = 0x555705078c80
        profileFolder = {static null = {<No data fields>}, d = 0x555705077280}
        myApp = {<QApplication> = {<No data fields>}, static staticMetaObject = {d = {superdata = {direct = 0x7ff280352fa0 <QApplication::staticMetaObject>}, stringdata = 0x7ff27e29b940 <qt_meta_stringdata_QgsApplication>, data = 0x7ff27e29bb40 <qt_meta_data_QgsApplication>, static_metacall = 0x7ff27d120194 <QgsApplication::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, static QGIS_ORGANIZATION_NAME = 0x7ff27e41dea2 "QGIS", static QGIS_ORGANIZATION_DOMAIN = 0x7ff27e41dea7 "qgis.org", static QGIS_APPLICATION_NAME = 0x7ff27e41deb0 "QGIS3", static settingsLocaleUserLocale = {<QgsSettingsEntryBase> = {_vptr.QgsSettingsEntryBase = 0x7ff27f07c8b0 <vtable for QgsSettingsEntryString+16>, mKey = {static null = {<No data fields>}, d = 0x7ff27e284fa0 <QgsApplication::{lambda()#1}::operator()() const::qstring_literal>}, mDefaultValue = {d = {data = {c = 96 '`', uc = 96 '`', s = -1696, sc = 96 '`', us = 63840, i = 2135423328, u = 2135423328, l = 140679494236512, ul = 140679494236512, b = 96, d = 6.9504905176582222e-310, f = 2.658112e+38, real = 6.9504905176582222e-310, ll = 140679494236512, ull = 140679494236512, o = 0x7ff27f47f960 <QArrayData::shared_null>, ptr = 0x7ff27f47f960 <QArrayData::shared_null>, shared = 0x7ff27f47f960 <QArrayData::shared_null>}, type = 10, is_shared = 0, is_null = 0}}, mSection = QgsSettings::NoSection, mDescription = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}, mPluginName = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}}, mMinLength = 0, mMaxLength = -1}, static settingsLocaleOverrideFlag = {<QgsSettingsEntryBase> = {_vptr.QgsSettingsEntryBase = 0x7ff27f07c840 <vtable for QgsSettingsEntryBool+16>, mKey = {static null = {<No data fields>}, d = 0x7ff27e284fe0 <QgsApplication::{lambda()#2}::operator()() const::qstring_literal>}, mDefaultValue = {d = {data = {c = 0 '\000', uc = 0 '\000', s = -23808, sc = 0 '\000', us = 41728, i = -1873894656, u = 2421072640, l = 140731319493376, ul = 140731319493376, b = false, d = 6.9530510255585623e-310, f = -4.07519073e-29, real = 6.9530510255585623e-310, ll = 140731319493376, ull = 140731319493376, o = 0x7ffe904ea300, ptr = 0x7ffe904ea300, shared = 0x7ffe904ea300}, type = 1, is_shared = 0, is_null = 0}}, mSection = QgsSettings::NoSection, mDescription = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}, mPluginName = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}}, <No data fields>}, static settingsLocaleGlobalLocale = {<QgsSettingsEntryBase> = {_vptr.QgsSettingsEntryBase = 0x7ff27f07c8b0 <vtable for QgsSettingsEntryString+16>, mKey = {static null = {<No data fields>}, d = 0x7ff27e285020 <QgsApplication::{lambda()#3}::operator()() const::qstring_literal>}, mDefaultValue = {d = {data = {c = 96 '`', uc = 96 '`', s = -1696, sc = 96 '`', us = 63840, i = 2135423328, u = 2135423328, l = 140679494236512, ul = 140679494236512, b = 96, d = 6.9504905176582222e-310, f = 2.658112e+38, real = 6.9504905176582222e-310, ll = 140679494236512, ull = 140679494236512, o = 0x7ff27f47f960 <QArrayData::shared_null>, ptr = 0x7ff27f47f960 <QArrayData::shared_null>, shared = 0x7ff27f47f960 <QArrayData::shared_null>}, type = 10, is_shared = 0, is_null = 0}}, mSection = QgsSettings::NoSection, mDescription = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}, mPluginName = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}}, mMinLength = 0, mMaxLength = -1}, static settingsLocaleShowGroupSeparator = {<QgsSettingsEntryBase> = {_vptr.QgsSettingsEntryBase = 0x7ff27f07c840 <vtable for QgsSettingsEntryBool+16>, mKey = {static null = {<No data fields>}, d = 0x7ff27e285060 <QgsApplication::{lambda()#4}::operator()() const::qstring_literal>}, mDefaultValue = {d = {data = {c = 0 '\000', uc = 0 '\000', s = -23808, sc = 0 '\000', us = 41728, i = -1873894656, u = 2421072640, l = 140731319493376, ul = 140731319493376, b = false, d = 6.9530510255585623e-310, f = -4.07519073e-29, real = 6.9530510255585623e-310, ll = 140731319493376, ull = 140731319493376, o = 0x7ffe904ea300, ptr = 0x7ffe904ea300, shared = 0x7ffe904ea300}, type = 1, is_shared = 0, is_null = 0}}, mSection = QgsSettings::NoSection, mDescription = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}, mPluginName = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}}, <No data fields>}, static settingsSearchPathsForSVG = {<QgsSettingsEntryBase> = {_vptr.QgsSettingsEntryBase = 0x7ff27f07c878 <vtable for QgsSettingsEntryStringList+16>, mKey = {static null = {<No data fields>}, d = 0x7ff27e2850c0 <QgsApplication::{lambda()#5}::operator()() const::qstring_literal>}, mDefaultValue = {d = {data = {c = 112 'p', uc = 112 'p', s = 1136, sc = 112 'p', us = 1136, i = 2135426160, u = 2135426160, l = 140679494239344, ul = 140679494239344, b = 112, d = 6.9504905177981416e-310, f = 2.6586864e+38, real = 6.9504905177981416e-310, ll = 140679494239344, ull = 140679494239344, o = 0x7ff27f480470 <QListData::shared_null>, ptr = 0x7ff27f480470 <QListData::shared_null>, shared = 0x7ff27f480470 <QListData::shared_null>}, type = 11, is_shared = 0, is_null = 0}}, mSection = QgsSettings::NoSection, mDescription = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}, mPluginName = {static null = {<No data fields>}, d = 0x7ff27f47f960 <QArrayData::shared_null>}}, <No data fields>}, static mFileOpenEventReceiver31900 = 0x555705219fd0, static mInitialized31900 = true, static mRunningFromBuildDir31900 = true, static sMaxThreads31900 = -1, mIconCache = {d = 0x555705074710}, mCursorCache = {d = 0x5557063d0760}, mQgisTranslator = 0x555705227340, mQtTranslator = 0x555705227190, mDataItemProviderRegistry = 0x555705f88b80, mAuthManager = 0x5557060373f0, mApplicationMembers = 0x7ff264006060, static sApplicationMembers = 0x0, static sAuthManager = 0x0}
        settings = {<QObject> = {<No data fields>}, static staticMetaObject = {d = {superdata = {direct = 0x7ff27f626100 <QObject::staticMetaObject>}, stringdata = 0x7ff27e2aa200 <qt_meta_stringdata_QgsSettings>, data = 0x7ff27e2aa240 <qt_meta_data_QgsSettings>, static_metacall = 0x7ff27d13287a <QgsSettings::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, mUserSettings = 0x5557050ab1d0, mGlobalSettings = 0x55570525c870, mUsingGlobalArray = false}
        libPaths = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = {<std::__atomic_base<int>> = {static _S_alignment = 4, _M_i = -1}, static is_always_lock_free = true}}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x555705acce60}, d = 0x555705acce60}}, <No data fields>}
        relLibPath = {static null = {<No data fields>}, d = 0x55570590c980}
        customizationsettings = 0x555705a67c40
        systemEnvVars = {d = 0x5557052756c0}
        useCustomVars = false
        desiredStyle = {static null = {<No data fields>}, d = 0x555705c330c0}
        theme = {static null = {<No data fields>}, d = 0x555705afee20}
        activeStyleName = {static null = {<No data fields>}, d = 0x555705b45290}
        mySplashPath = {static null = {<No data fields>}, d = 0x7ff27e41f0a0 <QgsApplication::splashPath()::{lambda()#1}::operator()() const::qstring_literal>}
        myPixmap = <incomplete type>
        w = 600
        h = 300
        mypSplash = 0x5557051e9c10
        qgis = 0x555705219fd0
        sigwatch = {<QObject> = {<No data fields>}, static staticMetaObject = {d = {superdata = {direct = 0x7ff27f626100 <QObject::staticMetaObject>}, stringdata = 0x555703245440 <qt_meta_stringdata_UnixSignalWatcher>, data = 0x555703245500 <qt_meta_data_UnixSignalWatcher>, static_metacall = 0x5557031ff79e <UnixSignalWatcher::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, d_ptr = 0x7ff264007680}
        retval = 52711808
[Inferior 1 (process 1407617) detached]
```

Doing the same in batches of 50 files handles the max open files limit more gracefully: it just refuses to open more files but does not crash.

**QGIS and OS versions**
02266ef8e6 on Debian bullseye

### PR DESCRIPTION
**Pull request by rouault**
I'm not sure if this is material OK for backport

### CODE REVIEW

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 1]
```diff
@@ -965,6 +966,19 @@ GDALDatasetH QgsOgrProviderUtils::GDALOpenWrapper( const char *pszPath, bool bUp
 {
   CPLErrorReset();
 
+  // Avoid getting too clause to the limit of files allowed in the process,
```
Comment (uclaros) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:969]:
```suggestion
  // Avoid getting too clοse to the limit of files allowed in the process,
```

File: src/core/qgsfileutils.h [hunk 2]
```diff
@@ -180,6 +180,40 @@ class CORE_EXPORT QgsFileUtils
      */
     static bool renameDataset( const QString &oldPath, const QString &newPath, QString &error SIP_OUT, Qgis::FileOperationFlags flags = Qgis::FileOperationFlag::IncludeMetadataFile | Qgis::FileOperationFlag::IncludeStyleFile );
 
+    /**
+     * Returns the limit of simultaneously opened file by the process.
```
Comment (uclaros) [inline src/core/qgsfileutils.h:184]:
```suggestion
     * Returns the limit of simultaneously opened files by the process.
```

File: src/core/qgsfileutils.h [hunk 3]
```diff
@@ -180,6 +180,40 @@ class CORE_EXPORT QgsFileUtils
...
+     * Currently only implemented on Unix.
+     *
+     * \returns Limit, or -1 if unknown.
+     * \note not available in Python bindings
+     * \since QGIS 3.22
+     */
+    static int openedFileLimit() SIP_SKIP;
+
+    /**
+     * Returns the number of currently opened file by the process.
```
Comment (uclaros) [inline src/core/qgsfileutils.h:195]:
```suggestion
     * Returns the number of currently opened files by the process.
```

File: src/core/qgsfileutils.h [hunk 4]
```diff
@@ -180,6 +180,40 @@ class CORE_EXPORT QgsFileUtils
...
+     *
+     * \returns Number of files, or -1 if unknown
+     * \note not available in Python bindings
+     * \since QGIS 3.22
+     */
+    static int openedFileCount() SIP_SKIP;
+
+    /**
+     * Returns whether when opening new file(s) will reach, or nearly reach,
+     * the limit of simultaneously opened file by the process.
```
Comment (uclaros) [inline src/core/qgsfileutils.h:207]:
```suggestion
     * the limit of simultaneously opened files by the process.
```

**Review by uclaros — COMMENTED** [review]
Nice!
Noted some comment typos.
How about hinting in the log that the user might want to raise `ulimit` to open all his files?

**Comment by rouault** [comment]
> How about hinting in the log that the user might want to raise `ulimit` to open all his files?

done

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@rouault 
> I'm not sure if this is material OK for backport

let's not. It's not a new issue, and I think it's ok to let it live on in 3.16.

**Pull request merged**